linera-witty/src/primitive_types/maybe_flat_type.rs

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! A representation of either a [`FlatType`] or nothing, represented by the unit (`()`) type.
+
+use super::flat_type::FlatType;
+use crate::memory_layout::{FlatLayout, Layout};
+use frunk::HCons;
+
+/// A marker trait for [`FlatType`]s and the unit type, which uses no storage space.
+pub trait MaybeFlatType: Sized {
+    /// Result of flattening the layout made up of the current element followed by `Tail`.
+    type Flatten<Tail: Layout>: FlatLayout;
+
+    /// Flattens a layout that starts with this [`MaybeFlatType`] followed by a provided `Tail`.
+    fn flatten<Tail>(self, tail: Tail) -> Self::Flatten<Tail>
+    where
+        Tail: Layout;
+}
+
+impl MaybeFlatType for () {
+    type Flatten<Tail: Layout> = Tail::Flat;
+
+    fn flatten<Tail>(self, tail: Tail) -> Self::Flatten<Tail>
+    where
+        Tail: Layout,
+    {
+        tail.flatten()
```

Comment:
Why does it matter to remove empty types from the list?

Reply:
The flat layout is mapped into the input parameters of a Wasm function (or the output results, but while Wasm supports multiple results types, WIT only supports one for now). There can't be any empty slots because there's no unit primitive in Wasm to use as a parameter, so we have to skip it. The generated code is simpler if the `FlatLayout` is known to never have any empty elements.